### PR TITLE
#461 FkContentType should have a Take rather than a Response.

### PR DESCRIPTION
**Pull request by dalifreire**
Issue #461 - FkContentType is built with a Response. To handle more general use cases like Content-Type: text/html; charset=..., it should be built with a Take that will handle the request dynamically.

### CODE REVIEW

**Comment by dalifreire** [comment]
@davvd Can you find someone to review it?

**Comment by davvd** [comment]
@dalifreire Thanks, someone will review your pull request soon

**Comment by davvd** [comment]
@mkordas it's yours, please review

**Comment by mkordas** [comment]
@dalifreire I'm on it

**Comment by mkordas** [comment]
@dalifreire please add issue number to PR description

**Comment by dalifreire** [comment]
@mkordas Thanks! Here we go again! Issue number added...

File: src/main/java/org/takes/facets/fork/FkContentType.java [hunk 1]
```diff
@@ -49,25 +50,25 @@
...
+    private final transient Take take;
 
     /**
      * Ctor.
      * @param atype Accepted type
-     * @param response Response to return
+     * @param take Take to handle the request dynamically.
      */
-    public FkContentType(final String atype, final Response response) {
+    public FkContentType(final String atype, final Take take) {
```
Comment (mkordas) [inline src/main/java/org/takes/facets/fork/FkContentType.java]: @dalifreire @HDouss are you guys sure that we should modify this constructor? It's in public API used in many projects, after update they won't compile anymore

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java [hunk 2]
```diff
@@ -59,14 +62,17 @@
...
-                    new RqFake(),
-                    FkContentTypeTest.CONTENT_TYPE,
-                    "*/* "
-                )
-            ).has(),
-            Matchers.is(true)
+            new FkContentType(
+                "text/xml",
+                FkContentTypeTest.createTakeWithEmptyResponse()
+                ).route(
```
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire there's something wrong with indentation here, this line should be 4 spaces back

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java [hunk 3]
```diff
@@ -78,8 +84,9 @@ public void matchesWithAnyTypes() throws IOException {
     public void matchesDifferentTypes() throws IOException {
         MatcherAssert.assertThat(
             new FkContentType(
-                "application/json charset=utf-8", new RsEmpty()
-            ).route(
+                "application/json charset=utf-8",
+                FkContentTypeTest.createTakeWithEmptyResponse()
+                ).route(
```
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire wrong indentation here as well

**Comment by mkordas** [comment]
@dalifreire see my comments above

File: src/main/java/org/takes/facets/fork/FkContentType.java (hunk 1 above)
Comment (HDouss) [inline src/main/java/org/takes/facets/fork/FkContentType.java]: @mkordas I am not sure FkContentType is used in many projects, especially how it was designed until now. Upgrading would be easy, wrapping with `TkFixed` will do the trick.

Comment (dalifreire) [inline src/main/java/org/takes/facets/fork/FkContentType.java]: @mkordas I'm not sure too...

**Comment by dalifreire** [comment]
@mkordas Fixed. Could you take a look again?

File: src/main/java/org/takes/facets/fork/FkContentType.java (hunk 1 above)
Comment (mkordas) [inline src/main/java/org/takes/facets/fork/FkContentType.java]: @HDouss you are right, it is not used

**Comment by mkordas** [comment]
@dalifreire checking

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java [hunk 4]
```diff
@@ -59,14 +62,17 @@
...
+                FkContentTypeTest.createTakeWithEmptyResponse()
+            ).route(
                 new RqWithHeader(
                     new RqFake(),
                     FkContentTypeTest.CONTENT_TYPE,
                     "*/* "
                 )
             ).has(),
-            Matchers.is(true)
+                Matchers.is(true)
```
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire indentation here was good, no need to change it

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java [hunk 5]
```diff
@@ -86,7 +93,7 @@ public void matchesDifferentTypes() throws IOException {
                     "images/*"
                 )
             ).has(),
-            Matchers.is(false)
+                Matchers.is(false)
```
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire same here

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java [hunk 6]
```diff
@@ -98,8 +105,9 @@ public void matchesDifferentTypes() throws IOException {
     public void matchesIdenticalTypes() throws IOException {
         MatcherAssert.assertThat(
             new FkContentType(
-                FkContentTypeTest.CTYPE, new RsEmpty()
-            ).route(
+                FkContentTypeTest.CTYPE,
+                FkContentTypeTest.createTakeWithEmptyResponse()
+                ).route(
```
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire please move this line 4 spaces back as well

**Comment by dalifreire** [comment]
@mkordas Please, check it again.

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java [hunk 7]
```diff
@@ -117,12 +125,14 @@ public void matchesIdenticalTypes() throws IOException {
...
-            new FkContentType("*/*", new RsEmpty()).route(
-                new RqWithHeader(
-                    new RqFake(), FkContentTypeTest.CONTENT_TYPE, ""
-                )
-            ).has(),
-            Matchers.is(true)
+            new FkContentType(
+                "*/*",
+                FkContentTypeTest.createTakeWithEmptyResponse()
+                ).route(
```
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire 4 spaces back

**Comment by dalifreire** [comment]
@mkordas Could you check if everything is right now?

**Comment by mkordas** [comment]
@dalifreire sure, I'm back

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java [hunk 8]
```diff
@@ -117,12 +125,14 @@ public void matchesIdenticalTypes() throws IOException {
...
-                new RqWithHeader(
-                    new RqFake(), FkContentTypeTest.CONTENT_TYPE, ""
-                )
-            ).has(),
-            Matchers.is(true)
+            new FkContentType(
+                "*/*",
+                FkContentTypeTest.createTakeWithEmptyResponse()
+            ).route(
+                    new RqWithHeader(
```
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire 
```
[ERROR] /src/test/java/org/takes/facets/fork/FkContentTypeTest.java[132]: Indentation (20) must be same or less than previous line (12), or bigger by exactly 4 (CascadeIndentationCheck)
```

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java (hunk 6 above)
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire please do it

**Comment by mkordas** [comment]
@dalifreire there are still 2 things to be fixed

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java (hunk 6 above)
Comment (dalifreire) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @mkordas What?

**Comment by dalifreire** [comment]
@mkordas Could you review again?

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java (hunk 6 above)
Comment (dalifreire) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @mkordas sorry, I see now

**Comment by mkordas** [comment]
@dalifreire looks good!

**Comment by mkordas** [comment]
@rultor merge

File: src/main/java/org/takes/facets/fork/FkContentType.java (hunk 1 above)
Comment (yegor256) [inline src/main/java/org/takes/facets/fork/FkContentType.java]: @dalifreire I don't think it's a good idea to remove the ctor. instead, add a new secondary one, which will use `TkFixed` inside

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java [hunk 9]
```diff
@@ -156,4 +167,17 @@ public void equalsAndHashCodeEqualTest() throws Exception {
             .suppress(Warning.TRANSIENT_FIELDS)
             .verify();
     }
+
+    /**
+     * Create a Take instance with empty response.
+     * @return Take
+     */
+    private static Take createTakeWithEmptyResponse() {
```
Comment (yegor256) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire read this: http://www.yegor256.com/images/books/elegant-objects/seven-pages.pdf

**Comment by yegor256** [comment]
@dalifreire two comments from me

**Comment by dalifreire** [comment]
@yegor256 Thanks for review! Please, take a look if it's better now...

File: src/main/java/org/takes/facets/fork/FkContentType.java [hunk 10]
```diff
@@ -59,15 +61,24 @@
      * @param response Response to return
      */
     public FkContentType(final String atype, final Response response) {
```
Comment (mkordas) [inline src/main/java/org/takes/facets/fork/FkContentType.java]: @dalifreire do we still have some code coverage for the old constructor?

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java [hunk 11]
```diff
@@ -59,7 +62,10 @@
     @Test
     public void matchesWithAnyTypes() throws IOException {
         MatcherAssert.assertThat(
-            new FkContentType("text/xml", new RsEmpty()).route(
+            new FkContentType(
+                "text/xml",
+                new RsEmpty()
+            ).route(
```
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire why do we need these changes? it was formatted OK

File: src/main/java/org/takes/facets/fork/FkContentType.java (hunk 10 above)
Comment (mkordas) [inline src/main/java/org/takes/facets/fork/FkContentType.java]: @dalifreire sorry, now I see we have

**Comment by mkordas** [comment]
@dalifreire see my comments

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java (hunk 11 above)
Comment (dalifreire) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @mkordas Some changes was requested by @yegor256. Take a look [here](https://github.com/dalifreire/takes/commit/e2ec249ca22ca9c85609c2349ea6304cc80cd948).

**Comment by dalifreire** [comment]
@mkordas Thanks. Indentation was fixed.

File: src/test/java/org/takes/facets/fork/FkContentTypeTest.java (hunk 11 above)
Comment (mkordas) [inline src/test/java/org/takes/facets/fork/FkContentTypeTest.java]: @dalifreire right, but here was just formatting change

**Comment by mkordas** [comment]
@dalifreire looks good!

**Comment by mkordas** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@elenavolokhova please, review this ticket for compliance with our [QA rules](http://at.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by elenavolokhova** [comment]
@davvd Looks good!

**Comment by davvd** [comment]
> @davvd Looks good!

@elenavolokhova thanks a lot :)

**Comment by davvd** [comment]
@mkordas added 10 mins to @elenavolokhova, for QA review, payment code is `79938921`

Many thanks! **40 mins** were added to your account in Transaction ID `56e145d60f6ed4019a000188` (task took 80 hours and 23 mins)

you're getting extra minutes here (c=25)

added +40 to your rating, now it is equal to [+8572](http://www.netbout.com/b/36080?open=rating)

**Pull request merged**